modape/modis/download.py

```diff
@@ -172,45 +177,76 @@ def _parse_response(query: List[dict]) -> dict:
+    def _parse_hdfxml(response):
+        result = {}
+        tree = ElementTree.fromstring(response.content)
+        for entry in tree.iter(tag='GranuleURMetaData'):
+            for datafile in entry.iter(tag="DataFiles"):
+                for datafilecont in datafile.iter(tag="DataFileContainer"):
+                    for content in datafilecont:
+                        if content.tag in ["Checksum", "FileSize"]:
+                            result.update({content.tag: int(content.text)})
+        return result
+
+
+    def _fetch(self,
+               session: SessionWithHeaderRedirection,
+               url: str,
+               destination: Path,
+               overwrite: bool,
+               check: bool,
+               ) -> Tuple[str, Union[None, Exception]]:
         """Helper function to fetch HDF files
 
         Args:
             session (SessionWithHeaderRedirection): requests session to fetch file.
             url (str): URL for file.
             destination (Path): Target directory.
             overwrite (bool): Overwrite existing.
+            check (bool): Check file size and checksum.
 
         Returns:
             Tuple[str, Union[None, Exception]]: Returns tuple with
                 either (filename, None) for success and (URL, Exception) for error.
 
         """
-
-        filename = destination.joinpath(url.split("/")[-1])
+        file_id = url.split("/")[-1]
+        filename = destination.joinpath(file_id)
 
         if not exists(filename) or overwrite:
 
+            filename_temp = filename.with_suffix(".modapedl")
+
             try:
 
                 with session.get(url, stream=True, allow_redirects=True) as response:
                     response.raise_for_status()
 
-                    with open(filename, "wb") as openfile:
+                    with open(filename_temp, "wb") as openfile:
                         shutil.copyfileobj(response.raw, openfile, length=16*1024*1024)#
 
-                assert filename.exists(), "File not on disk after download!"
+                if check:
+
+                    response = session.get(url + ".xml")
+                    response.raise_for_status()
+                    file_metadata = self._parse_hdfxml(response)
+
+                    # check filesize
+                    assert filename_temp.stat().st_size == file_metadata["FileSize"]
+                    with open(filename_temp, "rb") as openfile:
+                        checksum = cksum(openfile)
+                    # check checksum
+                    assert checksum == file_metadata["Checksum"]
 
-            except (HTTPError, AssertionError) as e:
-                return (url, e)
+                # QUESTION: should we remove temp file?
+                shutil.move(filename_temp, filename)
+
+            except (HTTPError, AssertionError, FileNotFoundError) as e:
+                return (file_id, e)
         else:
             log.info("%s exists in target. Please set overwrite to True.", filename)
```

Comment:
Good point, but thinking about it non-downloaded files should not be returned as downloaded. I understand this would cause problems for your applications, so we can leave as is and amend the documentation.

Reply:
Great. By the way, one could come up with arguments that it is just fine that *skipped* downloads are reported as if they were downloaded: (1) in the scenario of resuming a download query, it's nice that the final set of files on disk are reported as a single collection; (2) they are indeed available and do match the query -- so it's always about stuff that the user is indeed looking for. :))